gcp/api/server.py

```diff
@@ -149,6 +187,48 @@ def Watch(self, request, context: grpc.ServicerContext):
     context.abort(grpc.StatusCode.UNIMPLEMENTED, "Unimplemented")
 
 
+# Wrapped in a separate class
+@dataclass
+class ResponsesCount:
+  """Wraps responses count in a separate class 
+  to allow it to be passed by reference
+  
+  Also adds a interface to allow easy updating to a mutex
+  if necessary
+  """
+  count: int
+
+  def add(self, amount):
+    # This is to prevent query `limit` parameter being smaller than
+    # the number that is checked later in the iter() loop for the last page
+    if amount < 0:
+      raise ValueError("change amount must be positive")
+    self.count += amount
+
+  def exceeded(self) -> bool:
+    return self.count > _MAX_VULN_RESP_THRESH
+
+  def page_limit(self) -> int:
+    """
+    Returns the limit based on whether we have 
+    exceeded the _MAX_VULN_RESP_THRESH with the total number
+    of responses in the entire query batch
+    """
+    if self.exceeded():
+      return _MAX_VULN_LISTED_POST_EXCEEDED
+
+    return _MAX_VULN_LISTED_PRE_EXCEEDED
+
+
+@dataclass
+class QueryContext:
+  service_context: grpc.ServicerContext
+  page_token: ndb.Cursor | None
+  request_start_time: datetime
```

Comment:
Is this used anywhere?

Reply:
Commented this out, might still be useful in the future to add a hard time limit to avoid continuing the query after the 60s timeout.